theseus/theseus_layer.py

```diff
@@ -93,3 +118,154 @@ def device(self) -> torch.device:
+    """
+    Functionally the same as the forward method in a TheseusLayer
+    but computes the direct loss minimization in the backward pass.
+    """
+
+    _dlm_epsilon = "dlm_epsilon"
+    _grad_suffix = "_grad"
+
+    @staticmethod
+    def forward(
+        ctx,
+        objective,
+        optimizer,
+        optimizer_kwargs,
+        bwd_objective,
+        bwd_optimizer,
+        epsilon,
+        *input_data,
+    ):
+        n = len(input_data) // 2
+        input_keys = input_data[:n]
+        input_vals = input_data[n:]
+        ctx.n = n
+
+        input_data = dict(zip(input_keys, input_vals))
+        ctx.input_keys = input_keys
+
+        differentiable_keys = [k for k in input_keys if input_data[k].requires_grad]
+        differentiable_tensors = [v for v in input_vals if v.requires_grad]
+        ctx.differentiable_keys = differentiable_keys
+
+        optim_tensors, info = _forward(
+            objective, optimizer, optimizer_kwargs, input_data
+        )
+
+        # Skip computation if there are no differentiable inputs.
+        if len(differentiable_keys) > 0:
+            ctx.bwd_objective = bwd_objective
+            ctx.bwd_optimizer = bwd_optimizer
+            ctx.epsilon = epsilon
+
+            # Precompute and cache this.
+            with torch.enable_grad():
+                grad_sol = torch.autograd.grad(
+                    objective.error_squared_norm().sum(),
+                    differentiable_tensors,
+                    allow_unused=True,
+                )
+            ctx.k = len(differentiable_keys)
+            ctx.save_for_backward(*input_vals, *grad_sol, *optim_tensors)
+        return (*optim_tensors, info)
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, *grad_outputs):
+        saved_tensors = ctx.saved_tensors
+        input_vals = saved_tensors[: ctx.n]
+        grad_sol = saved_tensors[ctx.n : ctx.n + ctx.k]
+        optim_tensors = saved_tensors[ctx.n + ctx.k :]
+        grad_outputs = grad_outputs[:-1]
+
+        bwd_objective = ctx.bwd_objective
+        bwd_optimizer = ctx.bwd_optimizer
+        epsilon = ctx.epsilon
+        input_keys = ctx.input_keys
+        differentiable_keys = ctx.differentiable_keys
+
+        # Update the optim vars to their solutions.
+        bwd_data = dict(zip(input_keys, input_vals))
+        sol_values = dict(zip(bwd_objective.optim_vars.keys(), optim_tensors))
+        bwd_data.update(sol_values)
+
+        # Add in gradient values.
+        grad_data = {
+            TheseusLayerDLMForward._dlm_epsilon: torch.tensor(epsilon)
+            .to(grad_outputs[0])
+            .reshape(1, 1)
+        }
+        for i, name in enumerate(bwd_objective.optim_vars.keys()):
+            grad_data[name + TheseusLayerDLMForward._grad_suffix] = grad_outputs[i]
+        bwd_data.update(grad_data)
+
+        # Solve backward objective.
+        bwd_objective.update(bwd_data)
+        bwd_optimizer.optimize()
```

Comment:
Is there a chance here that the objective accidentally changes and the optimizer get's called on a stale version. Maybe using `bwd_optimizer.objective` is safer?